### ISSUE TITLE
Rectangle annotations don't target #xywh fragments

### ISSUE THREAD

**Issue by azaroth42** [ProjectMirador/mirador#1137]
Rectangle annotations are not produced as #xywh fragments, or with a FragmentSelector, as recommended by both the IIIF and Annotation specifications.  Instead they are SVG path elements that greatly reduce interoperability.

Note the strong warning not to do this at the very top of:  http://iiif.io/api/presentation/2.1/#non-rectangular-segments

**Comment by regisrob** [comment]
+1
I noticed the same limitation while working on a web page displaying rectangle annotations coming from Sesame/SAS and originally created in Mirador. The only way I found to recover the #xywh fragments from the <svg> and then set the IIIF Image URLs was to use SVG 1.1 [getBBox() function](https://www.w3.org/TR/SVG/types.html#__svg__SVGLocatable__getBBox) on the client side, which seems quite dirty... (there may be a clever way to do the same thing).

The tricky part on the Mirador side is that a rectangle annotation should become an SVG path as soon as it is rotated, I guess...

```
$("svg").each(function(){
        var path = this.querySelector("path");

        var rect = path.getBBox();
        var x = Math.ceil(rect.x);
        var y = Math.ceil(rect.y);
        var w = Math.ceil(rect.width);
        var h = Math.ceil(rect.height);
        var coord = x +","+ y +","+ w +","+ h;
...
```

<img width="1176" alt="capture d ecran 2016-09-27 a 09 20 07" src="https://cloud.githubusercontent.com/assets/402117/18864858/951a26c0-8498-11e6-9ae0-0122a3571829.png">

**Comment by dicksonlaw583** [comment]
I've noticed this as well. Even with non-rectangular or rotated-rectangular annotations, it is still useful to have a #xywh region for backwards compatibility with other viewers. Would it be possible to make Mirador use the SVG selector when it is present and the #xywh region as a backup when not?

**Comment by radpet** [comment]
@dicksonlaw583 I think that there is such support currently

Look at https://github.com/ProjectMirador/mirador/blob/master/js/src/annotations/osd-region-draw-tool.js#L94 if I am not wrong :)

**Comment by azaroth42** [comment]
I don't think that will ever get called. It would require the target of the annotation to be an object with a selector, and the xywh is the fragment part of the URI of the target, not as the value of a selector.

Unless it happens to be a FragmentSelector ... but I've never seen that used in the wild in IIIF.

See the "on" values  for the second two examples here:
    http://iiif.io/api/presentation/2.1/#segments

Hope that helps!

**Comment by glenrobson** [comment]
Just to note this is the type of annotation that came out of the previous version of Mirador:


```
{
  "@context" : "http://iiif.io/api/presentation/2/context.json",
  "@type" : "oa:Annotation",
  "motivation" : [ "oa:commenting" ],
  "resource" : [ {
    "@type" : "dctypes:Text",
    "format" : "text/html",
    "chars" : "<p>1852-1921</p>"
  } ],
  "on" : {
    "@type" : "oa:SpecificResource",
    "full" : "http://dams.llgc.org.uk/iiif/2.0/4574752/canvas/4574755.json",
    "within" : "http://dams.llgc.org.uk/iiif/4574752/manifest.json",
    "selector" : {
      "@type" : "oa:FragmentSelector",
      "value" : "xywh=2218,3217,371,78"
    }
  }
}
```
Includes a fragementselector by default.

**Comment by azaroth42** [comment]
Oh huh. Well, it's definitely a valid annotation :)  Would be great to go back to it for rectangles.

**Comment by dicksonlaw583** [comment]
This is how it should be done according to Presentation API 2.1: http://iiif.io/api/presentation/2.1/#non-rectangular-segments

**Comment by sdellis** [comment]
It seems the bounding box coordinates just need to be supplied no matter what.  Should be easy to get with the paperjs [bounds](http://paperjs.org/reference/shape/#bounds) method. I don't think it would hurt to supply both the svg and the media fragment no matter what the shape is.

**Comment by rsinghal** [comment]
That is a good point, regarding the `bounds` method.

**Comment by rsinghal** [comment]
@azaroth42 Can we include both selectors in the annotation, so something like this?

```
{
  "@context" : "http://iiif.io/api/presentation/2/context.json",
  "@type" : "oa:Annotation",
  "motivation" : [ "oa:commenting" ],
  "resource" : [ {
    "@type" : "dctypes:Text",
    "format" : "text/html",
    "chars" : "<p>1852-1921</p>"
  } ],
  "on" : {
    "@type" : "oa:SpecificResource",
    "full" : "http://dams.llgc.org.uk/iiif/2.0/4574752/canvas/4574755.json",
    "selector" : [
    {
      "@type" : "oa:FragmentSelector",
      "value" : "xywh=2218,3217,371,78"
    },
    {
      "@type":"oa:SvgSelector",
      "value": "<svg xmlns=\"...\"><path d=\"...\"/></svg>"
    }
    ]
  }
}
```
And then the client chooses which selector to use.

**Comment by azaroth42** [comment]
In the Web Annotation model, that's exactly how it would work :)
  https://www.w3.org/TR/annotation-model/#selectors

However IIIF 2.x uses the Open Annotation spec where it's more complicated. OA requires there be at most one Selector:
  http://openannotation.org/spec/core/specific.html#Selectors

And to have multiples you wrap the selectors in an oa:Choice:
  http://openannotation.org/spec/core/multiplicity.html#Choice

Thus ...

```
"on" : {
    "@type" : "oa:SpecificResource",
    "full" : "http://dams.llgc.org.uk/iiif/2.0/4574752/canvas/4574755.json",
    "selector" : {
      "@type": "oa:Choice",
      "default": {
        "@type" : "oa:FragmentSelector",
        "value" : "xywh=2218,3217,371,78"
      },
     "item": {
        "@type":"oa:SvgSelector",
        "value": "<svg xmlns=\"...\"><path d=\"...\"/></svg>"
      }
    }
  }
```

**Comment by dicksonlaw583** [comment]
I've looked at several other viewers that support annotations (Universal Viewer, Diva.js, Leaflet), and only the single-string "on" selector with #xywh has any real acceptance amongst them. Having multiple selector seems to be an area of the specifications that hardly anyone follows, we'll likely be the first.

**Comment by dfdan** [comment]
+1 for this issue - we have annotations (generated outside of Mirador) as xywh, that we display with Mirador. When delivered to Mirador as OADM they have oa:FragmentSelector selectors similar to what @glenrobson posted above. 

They work fine but they are converted to SVG paths (which I assume is required by paper.js). Probably doesn't matter how Mirador views them internally, but a tidy way to get xywh when saving them would be useful - and not just for non-rotated rectangular regions.

**Comment by rsinghal** [comment]
Fixed with #1222
